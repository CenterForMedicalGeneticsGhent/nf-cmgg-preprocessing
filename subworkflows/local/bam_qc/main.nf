#!/usr/bin/env nextflow

include { BAM_QC_PICARD         } from '../../nf-core/bam_qc_picard/main'
include { BAM_STATS_SAMTOOLS    } from "../../nf-core/bam_stats_samtools/main"

include { PICARD_BEDTOINTERVALLIST as BAITTOINTERVALLIST    } from '../../../modules/nf-core/picard/bedtointervallist/main'
include { PICARD_BEDTOINTERVALLIST as TARGETTOINTERVALLIST  } from '../../../modules/nf-core/picard/bedtointervallist/main'

workflow BAM_QC {
    take:
        ch_bam_bai          // channel: [mandatory] [meta, bam, bai]
        ch_fasta_fai        // channel: [mandatory] [meta2, fasta, fai]
        ch_fasta_dict       // channel: [mandatory][ meta2, dict ]
        ch_target_interval  // channel: [optional] [ target_interval_bed ]
        ch_bait_interval    // channel: [optional] [ bai_interval_bed ]
        disable_picard      // boolean: [optional] [ true ]

    main:
        ch_versions = Channel.empty()
        ch_metrics  = Channel.empty()

        ch_bait_interval_list = []
        ch_target_interval_list = []

        ch_fasta = ch_fasta_fai.map {meta, fasta, fai -> fasta             }.collect()
        ch_meta_fai   = ch_fasta_fai.map {meta, fasta, fai -> [meta, fai]  }.collect()
        ch_meta_fasta = ch_fasta_fai.map {meta, fasta, fai -> [meta, fasta]}.collect()

        if (!disable_picard) {
            if ( ch_bait_interval ) {
                BAITTOINTERVALLIST(
                    ch_bait_interval.map{it -> return [[id:"bait"], it ]},
                    ch_fasta_dict,
                    []
                )
                ch_versions = ch_versions.mix(BAITTOINTERVALLIST.out.versions)
                ch_bait_interval_list = BAITTOINTERVALLIST.out.interval_list.map{meta, list -> list}.collect()
            }
            if ( ch_target_interval ) {
                TARGETTOINTERVALLIST(
                    ch_target_interval.map{[[id:"target"], it ]},
                    ch_fasta_dict,
                    []
                )
                ch_versions = ch_versions.mix(TARGETTOINTERVALLIST.out.versions)
                ch_target_interval_list = TARGETTOINTERVALLIST.out.interval_list.map{meta, list -> list}.collect()
            }

            // SUBWORKFLOW: bam_qc_picard
            // Run Picard QC modules
            // BAM_QC_PICARD([meta, bam, bai], [meta2, fasta], [meta2, fai], [bait_interval], [target_interval])
            BAM_QC_PICARD( ch_bam_bai, ch_meta_fasta, ch_meta_fai, ch_bait_interval_list, ch_target_interval_list )
            ch_metrics  = ch_metrics.mix(BAM_QC_PICARD.out.coverage_metrics, BAM_QC_PICARD.out.multiple_metrics)
            ch_versions = ch_versions.mix(BAM_QC_PICARD.out.versions)
        }

        // SUBWORKFLOW: bam_stats_samtools
        // Run samtools QC modules
        // BAM_STATS_SAMTOOLS([meta, bam, bai])
        BAM_STATS_SAMTOOLS(ch_bam_bai, ch_fasta)

        ch_metrics = ch_metrics.mix(
            BAM_STATS_SAMTOOLS.out.stats,
            BAM_STATS_SAMTOOLS.out.flagstat,
            BAM_STATS_SAMTOOLS.out.idxstats
        ).groupTuple().map { meta, metrics -> [meta, metrics.flatten()] }
        ch_metrics.dump(tag: "BAM QC: metrics", {FormattingService.prettyFormat(it)})
        ch_versions = ch_versions.mix(BAM_STATS_SAMTOOLS.out.versions)

    emit:
        metrics  = ch_metrics   // [[meta, metrics], [...], ...]
        versions = ch_versions  // [versions]
}
