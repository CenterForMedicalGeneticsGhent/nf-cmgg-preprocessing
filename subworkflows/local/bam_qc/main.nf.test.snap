{
    "Bam QC - HSmetrics": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.stats:md5,7098b9422bf3cce350116e93bd8cf83a"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.flagstat:md5,167e69b479663a15194ddf56cbc9e60e"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.idxstats:md5,081d0431383fb7ea6b51b7077c6ec93c"
                ]
            ],
            "test.CollectMultipleMetrics.alignment_summary_metrics",
            "test.CollectHsMetrics.coverage_metrics"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-21T15:27:24.691191"
    },
    "Bam QC - Samtools": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.stats:md5,7098b9422bf3cce350116e93bd8cf83a"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.flagstat:md5,167e69b479663a15194ddf56cbc9e60e"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.idxstats:md5,081d0431383fb7ea6b51b7077c6ec93c"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-21T15:34:30.160724"
    },
    "Bam QC - WGSmetrics": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.stats:md5,7098b9422bf3cce350116e93bd8cf83a"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.flagstat:md5,167e69b479663a15194ddf56cbc9e60e"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.idxstats:md5,081d0431383fb7ea6b51b7077c6ec93c"
                ]
            ],
            "test.CollectMultipleMetrics.alignment_summary_metrics",
            "test.CollectWgsMetrics.coverage_metrics"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-21T15:28:20.716809"
    }
}